backend/service/service_proposal.go

```diff
@@ -170,6 +400,126 @@ func (service *ProposalService) QueryTypeAndTitleByIds(ids []uint64) ([]document
 	return proposalDocArr, err
 }
 
+func (_ ProposalService) GetValidatorPublicKeyMonikerFromProposalVoter(addrArrAsAa []string) (map[string]AddrAsMultiType, error) {
+
+	AddrAsMultiTypeMap := map[string]AddrAsMultiType{}
+
+	addrArrAsVa := make([]string, 0, len(addrArrAsAa))
+
+	for _, v := range addrArrAsAa {
+		va := utils.Convert(conf.Get().Hub.Prefix.ValAddr, v)
+		addrArrAsVa = append(addrArrAsVa, va)
+		AddrAsMultiTypeMap[v] = AddrAsMultiType{
+			AA: v,
+			Va: va,
+		}
+	}
+
+	var validatorsDoc []document.Validator
+	var selector = bson.M{"description.moniker": 1, "operator_address": 1, "consensus_pubkey": 1}
+
+	err := queryAll(document.CollectionNmValidator, selector, nil, "", 0, &validatorsDoc)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, validator := range validatorsDoc {
+
+		for k, v := range AddrAsMultiTypeMap {
+			if v.Va == validator.OperatorAddress {
+				v.ConsensusPubKey = validator.ConsensusPubkey
+				_, bytes, err := utils.DecodeAndConvert(validator.ConsensusPubkey)
+				if err != nil {
+					logger.Error("DecodeAndConvert", logger.String("err", err.Error()), logger.String("param", validator.ConsensusPubkey))
+					continue
+				}
+				v.ConsensusHex = strings.ToUpper(hex.EncodeToString(bytes))
+				AddrAsMultiTypeMap[k] = v
+			}
+		}
+	}
+
+	return AddrAsMultiTypeMap, nil
+}
+
+func (pro ProposalService) GetlVotingPowerForHeightArr(hArr []int64) (map[int64]votingPowerForHeight, error) {
+
+	var selector = bson.M{document.Block_Field_Height: 1, document.Block_Field_Validators: 1}
+
+	sort := desc(document.Block_Field_Height)
+	var blocks []document.Block
+	err := queryAll(document.CollectionNmBlock, selector, bson.M{"height": bson.M{"$in": hArr}}, sort, 0, &blocks)
+
+	if err != nil {
+		return nil, err
+	}
+
+	res := map[int64]votingPowerForHeight{}
+
+	for _, v := range blocks {
+		powerForHeight := int64(0)
+		voterPower := map[string]int64{}
+		for _, validator := range v.Validators {
+			powerForHeight += validator.VotingPower
+			voterPower[validator.PubKey] = validator.VotingPower
+		}
+		res[v.Height] = votingPowerForHeight{
+			totalVotingPower: powerForHeight,
+			voterPower:       voterPower,
+		}
+	}
+
+	return res, nil
+}
+
+func (_ ProposalService) GetMonikerByAddr(addrArr []string) (map[string]string, error) {
```

Comment:
consider this method can be replace by `GetValidatorPublicKeyMonikerFromProposalVoter `?